src/core/service.c

```diff
@@ -4908,6 +4911,38 @@ static void service_handoff_timestamp(
         unit_add_to_dbus_queue(u);
 }
 
+static int service_notify_pidref(Unit *u, PidRef *parent_pidref, PidRef *child_pidref) {
+        Service *s = ASSERT_PTR(SERVICE(u));
+        int r;
+
+        assert(parent_pidref);
+        assert(child_pidref);
+
+        if (pidref_equal(&s->main_pid, parent_pidref)) {
+                r = service_set_main_pidref(s, TAKE_PIDREF(*child_pidref), /* start_timestamp = */ NULL);
+                if (r < 0)
+                        return log_unit_warning_errno(u, r, "Failed to set new main pid: %m");
```

Comment:
Hmm, if this fails, the error would be propagated to manager_dispatch_..., and error in such handlers will result in the event source getting disabled. Please just make this a void, and always return 0 in handler.